### PR TITLE
Update clang from version 15 to 16

### LINKED ISSUE #37518: Update clang version on macOS to 16 to match conda-forge
**Describe the outcome that is desired.**
The `clang` version used in the `mantid-developer` conda environment on macOS increased to 16 to match conda-forge.

**Describe any solutions you are considering**
In `mantid-developer-osx.yml`, change 
```
- clang_osx-64=15
- clangxx_osx-64=15
```
to
```
- clang_osx-64=16
- clangxx_osx-64=16
```
When I tried this I had to remove my `mantid-developer` environment, then create it again, but then everything seemed to work.

**Additional context**
The version that conda-forge is using can be found [here](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml), at the time of writing this it is version 16. This was increased by conda-forge in October 2023 in [this](https://github.com/conda-forge/conda-forge-pinning-feedstock/commit/bac703af6f6a87e852a19f195caf60952c227061) commit. Probably there are also improvements from version 15 to 16 so another good reason to update.

### PR DESCRIPTION
**Pull request by jclarkeSTFC**
See #37518 for full details, but `conda-forge` uses clang version 16, so this updates our version to match. Similarly for `llvm_openmp`.

Fixed a compiler warning in `WideIntImpl.h` due to an implicit cast.

Fixes #37518. 

### To test (macOS-only):

- Remove your mantid-developer environment with `conda env remove -n mantid-developer`
- Delete `build/CMakeCache.txt`
- Create your mantid-developer environment with `mamba env create -f mantid-developer-osx.yml`
- Check that your environment has the right version by running `conda list | grep clang`
- Run `cmake --preset=osx`, the compiler version detected should be 16
- Compile
- Go and get a cup of tea
- Everything should compile and run as normal

Check that the macOS CI build used version 16 instead of 15 by looking at the build logs, they'll need a clean build to pick up the compiler change ~(I'm waiting until after 6.10 is released before triggering those)~.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Review by SilkeSchomann — DISMISSED** [review]
I have followed the steps and created mantid-developer with clang 16. A clean build of Mantid takes far too long to just drink a cup of tea, but is successful. Mantid also starts normally and behaved as expected.

The corresponding CI build for MacOS also used clang 16 and was successful

Happy to approve!

**Comment by sf1919** [comment]
@thomashampson can you gatekeep this one please?

File: docs/source/release/v6.11.0/Workbench/New_features/37518.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+- Updated compiler on macOS from clang version 15 to 16, which should resulti in performance improvements.
```
Comment (thomashampson) [inline docs/source/release/v6.11.0/Workbench/New_features/37518.rst:1]:
```suggestion
- Updated compiler on macOS from clang version 15 to 16, which should result in performance improvements.
```

**Review by SilkeSchomann — APPROVED** [review]
Approving again after a typo got fixed.

**Pull request merged**